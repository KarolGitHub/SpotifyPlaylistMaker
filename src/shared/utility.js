export const updateObject = (oldObject, updatedProps) => {
  return {
    ...oldObject,
    ...updatedProps,
  };
};

export const isValid = (value = "", rules = {}, isSignup = true) => {
  let isValid = true;

  if (!rules) {
    return true;
  }

  if (rules.required) {
    isValid = isSignup
      ? value.trim() !== "" &&
        (rules.minLength ? value.length >= rules.minLength : true) &&
        (rules.maxLength ? value.length <= rules.maxLength : true)
      : true;

    if (rules.isEmail) {
      const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
      isValid = pattern.test(value) && isValid;
    }

    if (rules.isNumeric) {
      const pattern = /^\d+$/;
      isValid = pattern.test(value) && isValid;
    }
    return isValid;
  }
};
