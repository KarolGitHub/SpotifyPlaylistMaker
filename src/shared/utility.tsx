export type SpotifyTrack = {
  id: string;
  name: string;
  artists: Array<{
    name: string;
  }>;
  album: {
    name: string;
  };
  year?: string;
  length?: string;
  uri?: string;
  preview_url?: string;
};

export type Track = {
  id: string;
  key?: number;
  name: string;
  artist: string;
  album: string;
  uri: string;
  preview_url: string;
};
export type Tracklist = Array<Track>;

// export type playStatus = 'PLAYING' | 'STOPPED' | 'PAUSED';

export type Tuple = [number, boolean, boolean];

export const updateObject = (oldObject: any, updatedProps?: any) => {
  return {
    ...oldObject,
    ...updatedProps,
  };
};

export const arrayDiff = (search: Tracklist, playlist: Tracklist) => [
  ...search.filter((searchItem) =>
    playlist.every((playlistItem) => searchItem.id !== playlistItem.id)
  ),
];

type Rules = {
  required?: boolean;
  minLength?: number;
  maxLength?: number;
  isEmail?: boolean;
  isNumeric?: boolean;
};

export const isValid = (value = "", rules: Rules) => {
  let isValid = true;

  if (!rules) {
    return true;
  }

  if (rules.required) {
    isValid =
      value.trim() !== "" &&
      (rules.minLength ? value.length >= rules.minLength : true) &&
      (rules.maxLength ? value.length <= rules.maxLength : true);
  }
  return isValid;
};

export const authPopup = (url: string) => {
  const width = 500;
  const height = 500;
  const left = window.screenX + (window.outerWidth - width) / 2;
  const top = window.screenY + (window.outerHeight - height) / 2.5;

  const windowFeatures = `toolbar=0,scrollbars=1,status=1,resizable=0,location=1,menuBar=0,width=${width},height=${height},top=${top},left=${left}`;

  const newWindow = window.open(url, "Spotify Playlist Maker", windowFeatures);
  newWindow?.focus();
  return newWindow;
};
