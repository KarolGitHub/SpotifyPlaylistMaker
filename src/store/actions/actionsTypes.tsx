export const ADD_TRACK = "ADD_TRACK";
export const DELETE_TRACK = "DELETE_TRACK";
export const SET_TRACKS = "SET_TRACKS";

export const SEARCH_TRACKS_START = "SEARCH_TRACKS_START";
export const SEARCH_TRACKS = "SEARCH_TRACKS";
export const SEARCH_TRACKS_SUCCESS = "SEARCH_TRACKS_SUCCESS";
export const SEARCH_TRACKS_FAIL = "SEARCH_TRACKS_FAIL";

export const SAVE_PLAYLIST_START = "SAVE_PLAYLIST_START";
export const SAVE_PLAYLIST = "SAVE_PLAYLIST";
export const SAVE_PLAYLIST_SUCCESS = "SAVE_PLAYLIST_SUCCESS";
export const SAVE_PLAYLIST_FAIL = "SAVE_PLAYLIST_FAIL";
export const SUCCESS_CONFIRM = "SUCCESS_CONFIRM";

export const EDIT_PLAYLIST_START = "EDIT_PLAYLIST_START";
export const EDIT_PLAYLIST = "EDIT_PLAYLIST";
export const EDIT_PLAYLIST_SUCCESS = "EDIT_PLAYLIST_SUCCESS";
export const EDIT_PLAYLIST_FAIL = "EDIT_PLAYLIST_FAIL";

export const AUTH_START = "AUTH_START";
export const AUTH_SUCCESS = "AUTH_SUCCESS";
export const AUTH_FAIL = "AUTH_FAIL";
export const AUTH_LOGOUT = "AUTH_LOGOUT";
export const SET_AUTH_REDIRECT_URL = "SET_AUTH_REDIRECT_URL";
export const AUTH_CHECK_STATE = "AUTH_CHECK_STATE";
export const AUTH_CHECK_TIMEOUT = "AUTH_CHECK_TIMEOUT";

export const PLAY_TRACK_STATE_UPDATE = "PLAY_TRACK_STATE_UPDATE";
export const PLAY_TRACK_START = "PLAY_TRACK_START";
export const PLAY_TRACK_PAUSE = "PLAY_TRACK_PAUSE";
export const PLAY_TRACK_END = "PLAY_TRACK_END";
export const PLAY_TRACK_FAIL = "PLAY_TRACK_FAIL";

export const FETCH_PLAYLISTS_START = "FETCH_PLAYLISTS_START";
export const FETCH_PLAYLISTS = "FETCH_PLAYLISTS";
export const FETCH_PLAYLISTS_SUCCESS = "FETCH_PLAYLISTS_SUCCESS";
export const FETCH_PLAYLISTS_FAIL = "FETCH_PLAYLISTS_FAIL";
export const CLEAR_PLAYLIST = "CLEAR_PLAYLIST";
export const SET_REDIRECT = "SET_REDIRECT";

export const FETCH_TRACKS_START = "FETCH_TRACKS_START";
export const FETCH_TRACKS = "FETCH_TRACKS";
export const FETCH_TRACKS_SUCCESS = "FETCH_TRACKS_SUCCESS";
export const FETCH_TRACKS_FAIL = "FETCH_TRACKS_FAIL";

export const DELETE_TRACKS_START = "DELETE_TRACKS_START";
export const DELETE_TRACKS = "DELETE_TRACKS";
export const DELETE_TRACKS_SUCCESS = "DELETE_TRACKS_SUCCESS";
export const DELETE_TRACKS_FAIL = "DELETE_TRACKS_FAIL";
