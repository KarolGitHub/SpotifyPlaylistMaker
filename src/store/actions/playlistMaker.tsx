import * as actionTypes from './actionsTypes';
import axios from '../../axios-spotify';
import { Dispatch } from 'redux';
import {
  Tracklist,
  updateObject,
  updateArray,
  PlaylistPayload,
} from '../../shared/utility';
import { AxiosResponse, AxiosError } from 'axios';

export const addTrack = (index: number) => {
  return {
    type: actionTypes.ADD_TRACK,
    index: index,
  };
};
export const deleteTrack = (index: number) => {
  return {
    type: actionTypes.DELETE_TRACK,
    index: index,
  };
};
export const moveTrack = (tracklists: {}) => {
  return {
    type: actionTypes.MOVE_TRACK,
    tracklists: tracklists,
  };
};

export const searchTracksStart = (limit: number) => {
  return {
    type: actionTypes.SEARCH_TRACKS_START,
    searchResultsLimit: limit,
  };
};
export const searchTracksSuccess = (tracklist: Tracklist) => {
  return {
    type: actionTypes.SEARCH_TRACKS_SUCCESS,
    tracklist: tracklist,
  };
};
export const searchTracksFail = (error: string) => {
  return {
    type: actionTypes.SEARCH_TRACKS_FAIL,
    error: error,
  };
};
export const searchTracks = (
  accessToken: string,
  country: string,
  searchType: string,
  term: string,
  limit: number
) => {
  return (dispatch: Dispatch) => {
    dispatch(searchTracksStart(limit));
    const queryParams = `search?type=${searchType}&limit=${limit}&q=${term}&market=${country}`;
    axios
      .get(queryParams, {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
      })
      .then((response) => JSON.parse(JSON.stringify(response)))
      .then((response) => {
        const tracks = updateArray(
          updateObject(response.data).tracks.items
        ).map((track: any) => ({
          id: track.id,
          name: track.name,
          artist: track.artists[0].name,
          album: track.album.name,
          uri: track.uri,
          preview_url: track.preview_url,
        }));
        dispatch(searchTracksSuccess(tracks));
      })
      .catch((error) => {
        if (error.response?.data) {
          dispatch(searchTracksFail(error.response.data.error.message));
        } else if (error.message) {
          dispatch(searchTracksFail(error.message));
        } else {
          dispatch(searchTracksFail('Unexpected Error!'));
        }
      });
  };
};

export const savePlaylistStart = () => {
  return {
    type: actionTypes.SAVE_PLAYLIST_START,
  };
};
export const savePlaylist = (
  accessToken: string,
  userId: string,
  userPayload: PlaylistPayload,
  trackURIs: Array<string>
) => {
  return (dispatch: Dispatch) => {
    dispatch(savePlaylistStart());
    const headers = {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
    };
    const queryParams = `users/${userId}/playlists`;
    axios
      .post(queryParams, JSON.stringify(userPayload), { headers: headers })
      .then((response: AxiosResponse) => JSON.parse(JSON.stringify(response)))
      .then((response: AxiosResponse) => {
        const playlistId = response.data.id;
        const queryParams = `users/${userId}/playlists/${playlistId}/tracks`;
        return axios.post(
          queryParams,
          JSON.stringify({
            uris: trackURIs,
          }),
          { headers: headers }
        );
      })
      .then(() => {
        dispatch(savePlaylistSuccess());
      })
      .catch((error: AxiosError) => {
        if (error.response?.data) {
          dispatch(savePlaylistFail(error.response.data.error.message));
        } else if (error.message) {
          dispatch(savePlaylistFail(error.message));
        } else {
          dispatch(savePlaylistFail('Unexpected Error!'));
        }
      });
  };
};
export const savePlaylistSuccess = () => {
  return {
    type: actionTypes.SAVE_PLAYLIST_SUCCESS,
  };
};
export const savePlaylistFail = (error: string) => {
  return {
    type: actionTypes.SAVE_PLAYLIST_FAIL,
    error: error,
  };
};

export const editPlaylistStart = () => {
  return {
    type: actionTypes.EDIT_PLAYLIST_START,
  };
};
export const editPlaylist = (
  accessToken: string,
  user_id: string,
  userPayload: PlaylistPayload,
  trackURIs: Array<string>,
  changedDetails: boolean,
  playlist_id?: string
) => {
  return (dispatch: Dispatch) => {
    dispatch(editPlaylistStart());
    const headers = {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
    };
    const queryParams = `users/${user_id}/playlists/${playlist_id}`;
    axios
      .get(queryParams, { headers: headers })
      .then(() => {
        if (changedDetails) {
          return axios.put(queryParams, JSON.stringify(userPayload), {
            headers: headers,
          });
        }
      })
      .then(() => {
        return axios.post(
          queryParams + '/tracks',
          JSON.stringify({
            uris: trackURIs,
          }),
          { headers: headers }
        );
      })
      .then(() => {
        dispatch(editPlaylistSuccess());
      })
      .catch((error: AxiosError) => {
        if (error.response?.data) {
          dispatch(editPlaylistFail(error.response.data.error.message));
        } else if (error.message) {
          dispatch(editPlaylistFail(error.message));
        } else {
          dispatch(editPlaylistFail('Unexpected Error!'));
        }
      });
  };
};
export const editPlaylistSuccess = () => {
  return {
    type: actionTypes.EDIT_PLAYLIST_SUCCESS,
  };
};
export const editPlaylistFail = (error: string) => {
  return {
    type: actionTypes.EDIT_PLAYLIST_FAIL,
    error: error,
  };
};

export const successConfirm = () => {
  return {
    type: actionTypes.SUCCESS_CONFIRM,
  };
};

export const setTracks = (tracklist: {}) => {
  return {
    type: actionTypes.SET_TRACKS,
    tracklist: tracklist,
  };
};
