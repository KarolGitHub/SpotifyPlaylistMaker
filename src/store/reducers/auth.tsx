import * as actionTypes from "../actions/actionsTypes";
import { updateObject } from "../../shared/utility";

const initialState = {
  token: null,
  userId: null,
  error: null,
  loading: false,
  authRedirectPath: "/",
};

const reducer = (
  state = initialState,
  action: {
    type: string;
    token: string;
    userId: string;
    error: string;
    path: string;
  }
) => {
  switch (action.type) {
    case actionTypes.AUTH_START:
      return updateObject(state, { error: null, loading: true });
    case actionTypes.AUTH_SUCCESS:
      return updateObject(state, {
        token: action.token,
        userId: action.userId,
        error: null,
        loading: false,
      });
    case actionTypes.AUTH_LOGOUT:
      return updateObject(state, { token: null, userId: null });
    case actionTypes.AUTH_REDIRECT:
      return updateObject(state, { authRedirectPath: action.path });
    case actionTypes.AUTH_FAIL:
      return updateObject(state, { error: action.error, loading: false });
    default:
      return state;
  }
};

export default reducer;
