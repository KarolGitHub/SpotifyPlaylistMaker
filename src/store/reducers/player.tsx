import * as actionTypes from "../actions/actionsTypes";
import { updateObject, Tuple } from "../../shared/utility";

export const initialState = {
  playerState: null,
  error: null,
};

const reducer = (
  state = initialState,
  action?: {
    type: string;
    playerState: Tuple | null | false;
    error: any;
  }
) => {
  switch (action?.type) {
    case actionTypes.PLAY_TRACK_STATE_UPDATE:
      return updateObject(state, { playerState: action.playerState });
    case actionTypes.PLAY_TRACK_START:
      return updateObject(state, { playerState: action.playerState });
    case actionTypes.PLAY_TRACK_PAUSE:
      return updateObject(state, { playerState: action.playerState });
    case actionTypes.PLAY_TRACK_END:
      return updateObject(state, { playerState: false });
    case actionTypes.PLAY_TRACK_FAIL:
      return updateObject(state, { error: action.error });
    default:
      return state;
  }
};

export default reducer;
