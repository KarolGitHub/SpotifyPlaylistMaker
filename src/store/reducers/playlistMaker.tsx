import * as actionTypes from "../actions/actionsTypes";
import {
  updateObject,
  Tracklist,
  tracksDiff,
  Track,
  spreadNestedObject,
} from "../../shared/utility";

type State = {
  playlist: Tracklist;
  searchResults: Tracklist;
  searchResultsLimit: number;
  error: any;
  loading: boolean;
  saveResult: boolean;
};
type Action = {
  type: string;
  index: number;
  track: Track;
  tracklist: Tracklist;
  searchResultsLimit: number;
  error: any;
  tracklists: {};
};
export const initialState: State = {
  playlist: [],
  searchResults: [],
  searchResultsLimit: 20,
  error: null,
  loading: false,
  saveResult: false,
};

const addTrack = (state: State, action: Action) => {
  return updateObject(state, {
    playlist: state.playlist.concat(state.searchResults[action.index]),
    searchResults: state.searchResults.filter(
      (_, index) => index !== action.index
    ),
  });
};

const deleteTrack = (state: State, action: Action) => {
  return updateObject(state, {
    playlist: state.playlist.filter((_, index) => index !== action.index),
  });
};
const moveTrack = (state: State, action: Action) => {
  return updateObject(state, spreadNestedObject(action.tracklists));
};

const searchTracksStart = (state: State, action: Action) => {
  return updateObject(state, {
    loading: true,
    searchResultsLimit: action.searchResultsLimit,
  });
};
const searchTracksSuccess = (state: State, action: Action) => {
  return updateObject(state, {
    searchResults: tracksDiff(action.tracklist, state.playlist),
    loading: false,
  });
};
const searchTracksFail = (state: State, action: Action) => {
  return updateObject(state, { error: action.error, loading: false });
};
const savePlaylistStart = (state: State) => {
  return updateObject(state, { loading: true });
};
const savePlaylistSuccess = (state: State) => {
  return updateObject(state, { loading: false, saveResult: true });
};
const savePlaylistFail = (state: State, action: Action) => {
  return updateObject(state, { error: action.error, loading: false });
};
const editPlaylistStart = (state: State) => {
  return updateObject(state, { loading: true });
};
const editPlaylistSuccess = (state: State) => {
  return updateObject(state, { loading: false, saveResult: true });
};
const editPlaylistFail = (state: State, action: Action) => {
  return updateObject(state, { error: action.error, loading: false });
};
const successConfirm = (state: State) => {
  return updateObject(state, { saveResult: false });
};
const setTracks = (state: State, action: Action) => {
  return updateObject(
    state,
    updateObject(spreadNestedObject(action.tracklist), {
      error: null,
    })
  );
};

const reducer = (state: State = initialState, action?: Action) => {
  switch (action?.type) {
    case actionTypes.ADD_TRACK:
      return addTrack(state, action);
    case actionTypes.DELETE_TRACK:
      return deleteTrack(state, action);
    case actionTypes.MOVE_TRACK:
      return moveTrack(state, action);
    case actionTypes.SEARCH_TRACKS_START:
      return searchTracksStart(state, action);
    case actionTypes.SEARCH_TRACKS_SUCCESS:
      return searchTracksSuccess(state, action);
    case actionTypes.SEARCH_TRACKS_FAIL:
      return searchTracksFail(state, action);
    case actionTypes.SAVE_PLAYLIST_START:
      return savePlaylistStart(state);
    case actionTypes.SAVE_PLAYLIST_SUCCESS:
      return savePlaylistSuccess(state);
    case actionTypes.SAVE_PLAYLIST_FAIL:
      return savePlaylistFail(state, action);
    case actionTypes.EDIT_PLAYLIST_START:
      return editPlaylistStart(state);
    case actionTypes.EDIT_PLAYLIST_SUCCESS:
      return editPlaylistSuccess(state);
    case actionTypes.EDIT_PLAYLIST_FAIL:
      return editPlaylistFail(state, action);
    case actionTypes.SUCCESS_CONFIRM:
      return successConfirm(state);
    case actionTypes.SET_TRACKS:
      return setTracks(state, action);
    default:
      return state;
  }
};

export default reducer;
