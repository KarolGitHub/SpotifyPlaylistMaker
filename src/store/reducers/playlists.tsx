import * as actionTypes from '../actions/actionsTypes';
import {
  updateObject,
  Playlist,
  Tracklist,
  updateArray,
  PlaylistInfo,
} from '../../shared/utility';

type State = {
  playlists: Array<Playlist>;
  tracks: Tracklist;
  error: any;
  loading: boolean;
  playlistInfo: PlaylistInfo | null;
  redirect: boolean;
};
type Action = {
  type: string;
  playlists: Array<Playlist>;
  tracks: Tracklist;
  error: any;
  playlistInfo: PlaylistInfo | null;
  redirect: boolean;
};
export const initialState: State = {
  playlists: [],
  tracks: [],
  error: null,
  loading: false,
  playlistInfo: null,
  redirect: false,
};
const fetchPlaylistsStart = (state: State) => {
  return updateObject(state, { loading: true });
};
const fetchPlaylistsSuccess = (state: State, action: Action) => {
  return updateObject(state, {
    playlists: updateArray(action.playlists),
    loading: false,
  });
};
const fetchPlaylistsFail = (state: State, action: Action) => {
  return updateObject(state, { loading: false, error: action.error });
};
const fetchTracksStart = (state: State) => {
  return updateObject(state, { loading: true });
};
const fetchTracksSuccess = (state: State, action: Action) => {
  return updateObject(state, {
    tracks: action.tracks,
    playlistInfo: action.playlistInfo,
    redirect: action.redirect,
    loading: false,
  });
};
const fetchTracksFail = (state: State, action: Action) => {
  return updateObject(state, { loading: false, error: action.error });
};
const deleteTracksStart = (state: State) => {
  return updateObject(state, { loading: true });
};
const deleteTracksSuccess = (state: State) => {
  return updateObject(state, { loading: false, playlists: [] });
};
const deleteTracksFail = (state: State, action: Action) => {
  return updateObject(state, { loading: false, error: action.error });
};
const clearPlaylist = (state: State) => {
  return updateObject(state, { playlistInfo: null, error: null, tracks: [] });
};
const setRedirect = (state: State) => {
  return updateObject(state, { redirect: false });
};

const reducer = (state: State = initialState, action?: Action) => {
  switch (action?.type) {
    case actionTypes.FETCH_PLAYLISTS_START:
      return fetchPlaylistsStart(state);
    case actionTypes.FETCH_PLAYLISTS_SUCCESS:
      return fetchPlaylistsSuccess(state, action);
    case actionTypes.FETCH_PLAYLISTS_FAIL:
      return fetchPlaylistsFail(state, action);
    case actionTypes.FETCH_TRACKS_START:
      return fetchTracksStart(state);
    case actionTypes.FETCH_TRACKS_SUCCESS:
      return fetchTracksSuccess(state, action);
    case actionTypes.FETCH_TRACKS_FAIL:
      return fetchTracksFail(state, action);
    case actionTypes.DELETE_TRACKS_START:
      return deleteTracksStart(state);
    case actionTypes.DELETE_TRACKS_SUCCESS:
      return deleteTracksSuccess(state);
    case actionTypes.DELETE_TRACKS_FAIL:
      return deleteTracksFail(state, action);
    case actionTypes.CLEAR_PLAYLIST:
      return clearPlaylist(state);
    case actionTypes.SET_REDIRECT:
      return setRedirect(state);
    default:
      return state;
  }
};

export default reducer;
