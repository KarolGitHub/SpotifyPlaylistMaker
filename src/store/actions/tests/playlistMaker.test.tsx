import configureMockStore from "redux-mock-store";
import * as actionTypes from "../actionsTypes";
import * as actions from "../index";
import reducer from "../../reducers/playlistMaker";
import { Tracklist, Track } from "../../../shared/utility";
import { initialState } from "../../reducers/playlistMaker";
const middlewares: any = [];
const mockStore = configureMockStore(middlewares);

describe("PlaylistMaker store", () => {
  const track: Track = {
    id: "id",
    name: "",
    album: "",
    artist: "",
    uri: "",
    preview_url: "",
  };
  const tracklist: Tracklist = [];
  describe("reducer should handle incoming actions", () => {
    it("return initial state on default", () => {
      expect(reducer()).toMatchSnapshot();
    });

    it("setTracks", () => {
      expect(
        reducer(initialState, {
          type: actionTypes.SET_TRACKS,
          index: 1,
          id: "id",
          tracklist: tracklist,
          searchResultsLimit: 10,
          error: null,
        })
      ).toMatchSnapshot();
    });
  });
  describe("action creators should dispatch actions", () => {
    let store: any;

    beforeEach(() => {
      store = mockStore(initialState);
    });

    it("setTracks", () => {
      store.dispatch(actions.setTracks(tracklist));
      expect(store.getActions()).toEqual([
        {
          type: actionTypes.SET_TRACKS,
          tracklist: [],
        },
      ]);
    });

    it("addTrack", () => {
      store.dispatch(actions.searchTracksSuccess([track]));
      store.dispatch(actions.addTrack(1, "id"));
      expect(store.getActions()).toEqual([
        {
          type: actionTypes.SEARCH_TRACKS_SUCCESS,
          tracklist: [
            {
              album: "",
              artist: "",
              id: "id",
              name: "",
              preview_url: "",
              uri: "",
            },
          ],
        },
        {
          type: actionTypes.ADD_TRACK,
          index: 1,
          id: "id",
        },
      ]);
      expect(store.getState()).toEqual({
        playlist: [],
        searchResults: [],
        searchResultsLimit: 20,
        addedTrackIDs: [],
        error: null,
        loading: false,
        saveResult: false,
      });
    });
  });
});
