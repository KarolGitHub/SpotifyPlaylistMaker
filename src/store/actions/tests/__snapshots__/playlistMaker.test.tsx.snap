// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistMaker store reducer should handle incoming actions return initial state on default 1`] = `
Object {
  "addedTrackIDs": Array [],
  "error": null,
  "loading": false,
  "playlist": Array [],
  "saveResult": false,
  "searchResults": Array [],
  "searchResultsLimit": 20,
}
`;

exports[`PlaylistMaker store reducer should handle incoming actions setTracks 1`] = `
Object {
  "addedTrackIDs": Array [],
  "error": null,
  "loading": false,
  "playlist": Array [],
  "saveResult": false,
  "searchResults": Array [],
  "searchResultsLimit": 20,
}
`;
